### PR TITLE
NIFI-6638: Empty multiple queues at once at different flow levels

### PR DESCRIPTION
**Pull request by VinceCastro**
This PR also refers to NIFI-3632, NIFI-5329 and NIFI-4308.

#### Description of PR
Replaced the current "Empty queue" button in the context menu with an "Empty queues" sub-menu providing different options to empty multiple queues at different levels of the flow depending on the selected components.

In particular, different buttons have been added to the new "Empty queues" sub-menu:
- selected queue: empty the selected queue
- selected queues: empty the selected queues
- current process group: empty all queues inside the current process group
- current process group (recursive): empty all queues inside the current process group recursively, which means all queues inside the current process group, and its sub process groups, and their sub process groups [and so on] will be emptied
- selected process group: empty all queues inside the selected process group
- selected process group (recursive): empty all queues inside the selected process group recursively
- selected process groups: empty all queues inside the selected process groups
- selected process groups (recursive): empty all queues inside the selected process groups recursively

**Errors management**: if the emptying process fails to empty one or more queues, it keeps going till all the selected queues are processed. After that, the final summary dialog eventually provides information about the errors faced in emptying one or more queues.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by adarmiento** [comment]
Are we sure we need all that many functionalities: 

> selected queue: empty the selected queue
> selected queues: empty the selected queues
> current process group: empty all queues inside the current process group
> current process group (recursive): empty all queues inside the current process group recursively, > which means all queues inside the current process group, and its sub process groups, and their sub process groups [and so on] will be emptied
> selected process group: empty all queues inside the selected process group
> selected process group (recursive): empty all queues inside the selected process group recursively
> selected process groups: empty all queues inside the selected process groups
> selected process groups (recursive): empty all queues inside the selected process groups recursively

Are there use cases in which you are interested into emptying all queue of a processor group which is not the current one AND at the same time we do not want to delete the queue recursively? 
I think that the most useful features are: 
> selected queue: empty the selected queue
> selected queues: empty the selected queues
>  current process group (recursive): empty all queues inside the current process group recursively

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 1]
```diff
@@ -212,6 +212,417 @@
             }).size() > 0;
     };
 
+    /**
+     * Empty all the queues inside the array of connections.
+     *
+     * @Param {string} actionName
+     * @param {Array} connections
+     * @Param {Array} errors
+     *
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]: nit: remove empty line
```suggestion
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/css/common-ui.css [hunk 2]
```diff
@@ -564,7 +564,7 @@ div.nifi-tooltip {
     background-color:rgba(249,250,251,0.97); /*tint base-color 96%*/
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/css/common-ui.css]:
```suggestion
    background-color: rgba(249,250,251,0.97); /*tint base-color 96%*/
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/css/common-ui.css [hunk 3]
```diff
@@ -564,7 +564,7 @@ div.nifi-tooltip {
...
     border:1px solid #004849; /*link-color*/
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/css/common-ui.css]:
```suggestion
    border: 1px solid #004849; /*link-color*/
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/css/common-ui.css [hunk 4]
```diff
@@ -564,7 +564,7 @@ div.nifi-tooltip {
...
     box-shadow: 0 3px 6px rgba(0,0,0,0.3);
-    width: 215px;
+    width: 235px;
     max-height: inherit;
     color:#004849
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/css/common-ui.css]:
```suggestion
    color: #004849;
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 5]
```diff
@@ -212,6 +212,417 @@
...
+
+        var createSuccessResponse = function () {
+            return {
+                success: true
+            };
+        };
+
+        // set the progress bar to a certain percentage
+        var setProgressBar = function (percentComplete) {
+            if($("#drop-request-percent-complete .progress-label").length) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
            if ($("#drop-request-percent-complete.progress-label").length) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 6]
```diff
@@ -212,6 +212,417 @@
...
+                $('#drop-request-percent-complete .md-hue-2 .md-container.md-mode-determinate .md-bar.md-bar2:first')
+                    .attr('style','transform: translateX(' + ((percentComplete - 100) / 2) + '%) scale(' + (percentComplete * 0.01) + ', 1);');
+
+                //set percentage
+                $("#drop-request-percent-complete .progress-label:first").text(percentComplete.toFixed(2) + '%');
+            }
+            else {
+                var progressBar = $('#drop-request-percent-complete');
+
+                (nfNgBridge.injector.get('$compile')($('<md-progress-linear ng-cloak ng-value="' + percentComplete + '" class="md-hue-2" md-mode="determinate" aria-label="Drop request percent complete"></md-progress-linear>'))(nfNgBridge.rootScope)).appendTo(progressBar);
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]: please split this line into multiple lines

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 7]
```diff
@@ -212,6 +212,417 @@
...
+            dropRequests.forEach(function (dropRequest) {
+                if( nfCommon.isDefinedAndNotNull(dropRequest) &&
+                    nfCommon.isDefinedAndNotNull(dropRequest.droppedCount) &&
+                    nfCommon.isDefinedAndNotNull(dropRequest.originalCount) ) {
+                    droppedFlowfiles += dropRequest.droppedCount;
+                    totalFlowfilesToDrop += dropRequest.originalCount;
+                }
+            });
+
+            if(totalFlowfilesToDrop !== 0) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
            if (totalFlowfilesToDrop !== 0) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 8]
```diff
@@ -212,6 +212,417 @@
...
+            }
+
+            //hide the status dialog
+            $('#drop-request-status-dialog').modal('hide');
+
+            var droppedFlowfiles = 0;
+            var droppedFlowfilesSize = 0;
+            var errorMessages = "";
+
+            if(nfCommon.isDefinedAndNotNull(errors)) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
            if (nfCommon.isDefinedAndNotNull(errors)) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 9]
```diff
@@ -212,6 +212,417 @@
...
+                errors.forEach(function (message) {
+                    errorMessages += '<p>ProcessGroupID ' + message.processGroupId + ': <span style="color: red">' +  nfCommon.escapeHtml(message.errorMessage) + '</span></p>';
+                });
+            }
+
+            //check for 403 error
+            for(var i = 0; i < responses.length; i++) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
            for (var i = 0; i < responses.length; i++) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 10]
```diff
@@ -212,6 +212,417 @@
...
+                var response = responses[i];
+                if(response.success === false && response.xhr.status === 403) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
                if (response.success === false && response.xhr.status === 403) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 11]
```diff
@@ -212,6 +212,417 @@
...
+                    errorMessages += '<p>QueueID ' + connections[i].id + ': <span style="color: red">' + response.xhr.status + ' - ' + nfCommon.escapeHtml(response.xhr.responseText) + '</span></p>';
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]: please split this line

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 12]
```diff
@@ -212,6 +212,417 @@
...
+                }
+            }
+
+            dropRequests.forEach(function (dropRequest) {
+                if(nfCommon.isDefinedAndNotNull(dropRequest)) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
                if (nfCommon.isDefinedAndNotNull(dropRequest)) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 13]
```diff
@@ -212,6 +212,417 @@
...
+                    // if this request failed show the error
+                    if (nfCommon.isDefinedAndNotNull(dropRequest.failureReason)) {
+                        errorMessages += '<p>QueueID ' + dropRequest.id + ': <span style="color: red">' + dropRequest.failureReason + '</span></p>';
+                    }
+                }
+            });
+
+            var results = $('<div></div>');
+
+            if(droppedFlowfiles !== 0) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
            if (droppedFlowfiles !== 0) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 14]
```diff
@@ -212,6 +212,417 @@
...
+                $('<span class="label"></span>').text(droppedFlowfiles).appendTo(results);
+                $('<span></span>').text(' FlowFiles (' + nfCommon.toReadableBytes(droppedFlowfilesSize) + ')').appendTo(results);
+                $('<span></span>').text(' were removed from the ' + (connections.length > 1 ? 'queues.' : 'queue.' )).appendTo(results);
+            }
+            else {
+                results.text('No FlowFiles were removed.');
+            }
+
+            if(errorMessages !== "") {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
            if (errorMessages !== "") {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 15]
```diff
@@ -212,6 +212,417 @@
...
+            deferredConnectionsResponse.resolve({
+                success: false,
+                xhr: xhr,
+                status: status,
+                error: error
+            });
+        });
+
+        // get process groups if recursive
+        if(recursive) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
        if (recursive) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 16]
```diff
@@ -212,6 +212,417 @@
...
+        }
+
+        $.when(deferredConnectionsResponse, deferredProcessGroupsResponse)
+            .done(function (connectionsResponse, processGroupsResponse) {
+                var response = {
+                    connections: [],
+                    errorMessages: []
+                };
+
+                if(connectionsResponse.success) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
                if (connectionsResponse.success) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 17]
```diff
@@ -212,6 +212,417 @@
...
+                    response.connections = connectionsResponse.response;
+                }
+                else {
+                    response.errorMessages.push({
+                        processGroupId: processGroupId,
+                        errorMessage: connectionsResponse.xhr.status + ' - Unable to get queues. ' + connectionsResponse.xhr.responseText
+                    });
+                }
+
+                if(!recursive) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
                if (!recursive) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 18]
```diff
@@ -212,6 +212,417 @@
...
+                    deferredResponse.resolve(response);
+                }
+                else {
+                    if(!processGroupsResponse.success) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
                    if (!processGroupsResponse.success) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 19]
```diff
@@ -212,6 +212,417 @@
...
+                    }
+                    else {
+                        var requests = processGroupsResponse.response.map(function (processGroup) {
+                            return getProcessGroupConnections(processGroup.id,true);
+                        });
+
+                        $.when.apply($,requests).then(function () {
+                            var responses = arguments;
+
+                            for(var i = 0; i < responses.length; i++) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
                            for (var i = 0; i < responses.length; i++) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 20]
```diff
@@ -212,6 +212,417 @@
...
+
+        $.when.apply($,deferredResponses).then(function () {
+            var responses = arguments;
+
+            var response = {
+                connections: [],
+                errorMessages: []
+            };
+
+            for(var i = 0; i < responses.length; i++) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
            for (var i = 0; i < responses.length; i++) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 21]
```diff
@@ -1086,183 +1497,173 @@
...
+            var dialogContent = selection.size() > 1
+                ? 'Are you sure you want to empty the selected queues? All FlowFiles waiting at the time of the request will be removed.'
+                : 'Are you sure you want to empty the selected queue? All FlowFiles waiting at the time of the request will be removed.';
 
             // prompt the user before emptying the queue
             nfDialog.showYesNoDialog({
-                headerText: 'Empty Queue',
-                dialogContent: 'Are you sure you want to empty this queue? All FlowFiles waiting at the time of the request will be removed.',
+                headerText: actionName,
+                dialogContent: dialogContent,
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]: Are you sure about this?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 22]
```diff
@@ -1086,183 +1497,173 @@
...
+         * @param {type} selection
+         */
+        emptyProcessGroupsQueues: function (selection) {
+            var selectionSize = selection.size();
+            var connections = [];
+            var errors = [];
+            var actionName = '';
+            var dialogContent = '';
+
+            if(selectionSize === 0) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
            if (selectionSize === 0) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 23]
```diff
@@ -1086,183 +1497,173 @@
...
+                if(connections.length === 0) {
+                    // display the "no queues to empty" dialog
+                    nfDialog.showOkDialog({
+                        headerText: actionName,
+                        dialogContent: 'No queues to empty.'
+                    });
+                    return;
+                }
+            }
+            else if(selectionSize === 1) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
            else if (selectionSize === 1) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 24]
```diff
@@ -1086,183 +1497,173 @@
...
-                        // reload the connection status
-                        nfConnection.reloadStatus(connection.id);
+            // prompt the user before emptying the queue
+            nfDialog.showYesNoDialog({
+                headerText: actionName,
+                dialogContent: dialogContent,
+                noText: 'Cancel',
+                yesText: 'Empty',
+                yesHandler: function () {
+                    if(selectionSize === 0) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
                    if (selectionSize === 0) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 25]
```diff
@@ -1086,183 +1497,173 @@
...
-                                // if this request failed so the error
-                                if (nfCommon.isDefinedAndNotNull(dropRequest.failureReason)) {
-                                    $('<br/><br/><span></span>').text(dropRequest.failureReason).appendTo(results);
-                                }
+                            response.errorMessages.forEach(function (errorMessage) {
+                                errors.push(errorMessage);
+                            });
 
-                                // display the results
+                            if(connections.length === 0 && errors.length === 0) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
                            if (connections.length === 0 && errors.length === 0) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 26]
```diff
@@ -1086,183 +1497,173 @@
...
-                        // update the percent complete
-                        updateProgress(dropRequest.percentCompleted);
+        emptyProcessGroupsQueuesRecursive: function (selection) {
+            var selectionSize = selection.size();
+            var connections = [];
+            var errors = [];
+            var actionName = '';
+            var dialogContent = '';
+
+            if(selectionSize === 0) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
            if (selectionSize === 0) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 27]
```diff
@@ -1086,183 +1497,173 @@
...
+                actionName = 'Empty Current Process Group Queues (Recursive)';
+                dialogContent = 'Are you sure you want to empty all queues inside the current process group and all its sub process groups (recursive)? All FlowFiles waiting at the time of the request will be removed.';
+                connections = d3.selectAll('g.connection').data();
+            }
+            else if(selectionSize === 1) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
            else if (selectionSize === 1) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 28]
```diff
@@ -1086,183 +1497,173 @@
...
+                yesText: 'Empty',
+                yesHandler: function () {
+                    var processGroupIDs = selectionSize === 0
+                        ?
+                        d3.selectAll('g.process-group')
+                            .data()
+                            .map(function (processGroup) {
+                                return processGroup.id;
+                            })
+                        :
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]: nit

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 29]
```diff
@@ -1086,183 +1497,173 @@
...
-
-                        // process the drop request
-                        dropRequest = response.dropRequest;
-                        processDropRequest(1);
-                    }).fail(function (xhr, status, error) {
-                        if (xhr.status === 403) {
-                            nfErrorHandler.handleAjaxError(xhr, status, error);
-                        } else {
-                            completeDropRequest()
+                        if(connections.length === 0 && errors.length === 0) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]:
```suggestion
                        if (connections.length === 0 && errors.length === 0) {
```

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/nf-common.js [hunk 30]
```diff
@@ -1390,6 +1390,24 @@
...
+        /**
+         * Converts a number of bytes into a readable format e.g KB, MB, GB, TB, YB
+         *
+         * @param {Int} bytes The number of bytes
+         * @returns {String}
+         */
+        toReadableBytes: function (bytes) {
+            var sizes = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
+
+            if(bytes === 0) {
```
Comment (adarmiento) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/nf-common.js]:
```suggestion
            if (bytes === 0) {
```

**Comment by adarmiento** [comment]
Some minor style fixes in the suggestions

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js (hunk 21 above)
Comment (VinceCastro) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]: Yep! Now the `headerText` and `dialogContent` are initialized depending on the specific "empty queue" menu item the user clicked on

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js (hunk 28 above)
Comment (VinceCastro) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]: Would you rather prefer the following?

```
var processGroupIDs;

if (selectionSize === 0) {
   processGroupIDs = ...
}
else {
   processGroupIDs = ...
}

```

**Comment by mcgilman** [comment]
@VinceCastro Thanks for the PR! This capability is important and is something that many folks have requested. The proposed solution will perform an unbounded number of requests in order to clear all nested queues. I think we need to change this as this solution likely won't scale well with larger data flows.

When the subject to the action is a process group, we should issue a single request to the backend with a flag indicating if the request is recursive. The clear queue action is an asynchronous request that the UI polls while the action completes.  When the action encompasses multiple queues we should be able to provide this context with the same asynchronous request/response model.

**Comment by VinceCastro** [comment]
Got it! So you suggest to move the empty multiple queues logic from the front-end to the back-end side by adding one or more rest api to the exsisting endpoints. We may have cases where the user wants to empty different selected process groups at once, or different selected queues at once, so I would suggest to add two rest api to the "FlowFile Queues" endpoint: the first one should accept a list of process group IDs in order to empty the queues inside them(recursively or not); the second one should accept a list of connections IDs to empty. We may collect all IDs into a json body.

This way we can efficiently cover all cases where the user wants to:

- empty the selected queue
- empty the selected queues
- empty the selected process group (recursively or not)
- empty the selected process groups (recursively or not)
- empty the current process group (recursively or not)

What do you think about?

@adarmiento you stated:

> I think that the most useful features are:
> 
>> selected queue: empty the selected queue
>> selected queues: empty the selected queues
>> current process group (recursive): empty all queues inside the current process group recursively

What if I want to empty only a subset of the flow I'm currently viewing? In that case features like "empty the selected process group" or "empty the selected process groups" are necessary, do you agree?

However, as mentioned by @adarmiento
> Are there use cases in which you are interested into emptying all queue of a processor group which is not the current one AND at the same time we do not want to delete the queue recursively?

I have doubts about the usefulness of the non recursive actions. Are there use cases where we do not really need to empty a process group recursively? Do you guys @mcgilman @adarmiento suggest to add the non recursive actions or postpone them to another pr?

**Comment by adarmiento** [comment]
I'd say that to empty a subset of the flow I am currently viewing, I'd go for the "selected queues: empty the selected queues", however, I understand your urge to add more convenience methods.
I'd say I would add all the non-core feature in 1+ separate PRs, keeping this "core" implementation as simple as possible.

**Comment by patricker** [comment]
@VinceCastro I'd leave the choice to keep it as a single PR, or split it, up to your comfort level with the back-end code.  This is your first contribution, have you looked at the back-end layer?

A few file pointers in case you want to look into this:
 - The main end point is in nifi-web-api/org.apache.nifi.web.api.FlowFileQueueResource
 - Do a quick search for `@Path("{id}/drop-requests")` and you'll find the end point.

This is a POST for a single Queue.  The ID of the queue goes in the URL.  You might be able to replace this with the ID of a process group and then do a lookup, but probably easier to create a new end point for `@Path("{id}/process-group-drop-requests`.

But what about submitting multiple queues async?  I think you would need another endpoint there too, and have the body of the post contain some JSON describing the list.

Interestingly, the `DropRequestDTO` does not specify which Queue it's for, so you wouldn't need to worry about that part too much.  The harder part is probably the actual drop logic in NiFi, and updating the other internal objects like `StandardNiFiServiceFacade.createFlowFileDropRequest`.

**Comment by VinceCastro** [comment]
Hi and thanks for the useful tips! I'm currently checking the back-end out, I'm going to decide later whether or not to split the pr.

**Comment by VinceCastro** [comment]
I was thinking about remote process groups, should we take care of them some way or just ignore them during the emptying process?

**Comment by VinceCastro** [comment]
@adarmiento @mcgilman @patricker I moved the multiple queues emptying logic from the front-end to the back-end side, could you please check the changes out?

**Comment by adarmiento** [comment]
Could you please add some more test cases enforcing your expected behavior?

**Comment by joewitt** [comment]
closing due to inactivity.  but looks like it was a great effort/contrib.  hopefully restored at some point

**Pull request closed without merging**